cmd/argo/commands/cp.go

```diff
@@ -0,0 +1,143 @@
+package commands
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/argoproj/argo-workflows/v3/cmd/argo/commands/client"
+	"github.com/argoproj/argo-workflows/v3/pkg/apiclient"
+	workflowpkg "github.com/argoproj/argo-workflows/v3/pkg/apiclient/workflow"
+	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+)
+
+type copyArtifactOpts struct {
+	namespace    string // --namespace
+	nodeId       string // --node-id
+	templateName string // --template-name
+	artifactName string // --artifact-name
+	path         string // --path
+}
+
+func NewCpCommand() *cobra.Command {
+	var cpArtifactOpts copyArtifactOpts
+
+	command := &cobra.Command{
+		Use:   "cp my-wf output-directory ...",
+		Short: "copy artifacts from workflow",
+		Example: `# Copy a workflow's artifacts to a local output directory:
+
+  argo cp my-wf output-directory
+
+# Copy artifacts from a specific node in a workflow to a local output directory:
+
+  argo cp my-wf output-directory --node-id=my-wf-node-id-123
+`,
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 2 {
+				cmd.HelpFunc()(cmd, args)
+				return fmt.Errorf("incorrect number of arguments")
+			}
+			workflowName := args[0]
+			outputDir := args[1]
+
+			ctx, apiClient := client.NewAPIClient(cmd.Context())
+			serviceClient := apiClient.NewWorkflowServiceClient()
+			namespace := client.Namespace()
+			if len(cpArtifactOpts.namespace) > 0 {
+				namespace = cpArtifactOpts.namespace
+			}
+			workflow, err := serviceClient.GetWorkflow(ctx, &workflowpkg.WorkflowGetRequest{
+				Name:      workflowName,
+				Namespace: namespace,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to get workflow: %w", err)
+			}
+
+			workflowName = workflow.Name
+			artifactSearchQuery := v1alpha1.ArtifactSearchQuery{
+				ArtifactName: cpArtifactOpts.artifactName,
+				TemplateName: cpArtifactOpts.templateName,
+				NodeId:       cpArtifactOpts.nodeId,
+			}
+			artifactSearchResults := workflow.SearchArtifacts(&artifactSearchQuery)
+
+			c := &http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: client.ArgoServerOpts.InsecureSkipVerify,
+					},
+				},
+			}
+
+			for _, artifact := range artifactSearchResults {
+				customPath := filepath.Join(outputDir, cpArtifactOpts.path)
+				nodeInfo := workflow.Status.Nodes.Find(func(n v1alpha1.NodeStatus) bool { return n.ID == artifact.NodeID })
+				if nodeInfo != nil {
```

Comment:
if `nodeInfo == nil` that's an error surely?

Reply:
That's true - that would mean the user entered the wrong template name. I'll update this.

Reply:
Updated to throw an error if nodeInfo is `nil`